### ISSUE TITLE
paste-retaining-clones can corrupt the outline

### ISSUE THREAD

**Issue by SegundoBob** [leo-editor/leo-editor#862]
### Overview

If I try at one time to move several nodes while preserving their gnx's, I cause this traceback:

```
Traceback (most recent call last):

  File "/pri/git/leo-editor/leo/core/leoCommands.py", line 2016, in executeAnyCommand
    return command(event)

  File "/pri/git/leo-editor/leo/core/leoGlobals.py", line 2445, in commander_command_wrapper
    method(event=event)

  File "/pri/git/leo-editor/leo/commands/commanderOutlineCommands.py", line 1390, in promote
    p.promote()

  File "/pri/git/leo-editor/leo/core/leoNodes.py", line 1596, in promote
    child.parents.remove(p.v)

ValueError: list.remove(x): x not in list

```
The Leo-Editor session is left corrupted.  That is, it continues to behave badly till the file is closed and reopened.

## Detailed Procedure for Reproducing the Problem

The same problem occurs when I only move one node with the procedure that would move any number of nodes.  So the following procedure reproduces the problem:

1. Open a new Leo-Editor file.

1. Ctrl-I (create a second root node)

1. Ctrl-I (create a third root node)

1. Ctrl-I (create a fourth root node)

1. Alt-UpArrow (move the focus to the third root node)

1. Ctrl-] (make the fourth root node a child of the third root node)

1. Shift-Ctrl-X (Delete the subtree rooted by third root node)

1. Alt-UpArrow (move the focus to the first root node)

1. Shift-Alt-V (Paste retaining gnx's)

1. Ctrl-[ (promote all the children of the just pasted node.  This causes the traceback.)

### Important Observations

There is no error if a Shift-Ctrl-V (paste with new gnx's) is used in the above procedure.

The "paste retaining gnx's" applies to all the nodes pasted, not just the root of the subtree pasted.  So in order to do what I want to do (that is retain the gnx's of the children) I must use Shift-Alt-V (paste retaining gnx's).

Perhaps the root of this problem is that Shift-Ctrl-X fails to clear the parent list of the deleted node?

### Test System
```
Xubuntu 16.04
Leo 5.7.1, build 20180402204423, Mon Apr  2 20:44:23 CDT 2018
Git repo info: branch = Leo-Babel, commit = 0be201e4a762
Python 3.5.2, PyQt version 5.5.1
linux
isPython3: True
caching enabled
```

**Comment by edreamleo** [comment]
Bob, what happens if you set g.SQLITE to False?

**Comment by SegundoBob** [comment]
I get the same error when g.SQLITE is False.

I use Ctrl-B to set g.SQLITE to False before beginning the 10 step bug replication.
After the error g.SQLITE is still False.

**Comment by edreamleo** [comment]
On Thu, Apr 19, 2018 at 11:08 AM, SegundoBob <notifications@github.com>
wrote:

> I get the same error when g.SQLITE is False.
>
> I use Ctrl-B to set g.SQLITE to False before beginning the 10 step bug
> replication.
> After the error g.SQLITE is still False.
>
​Thanks for this.  It's good to know.

Edward

**Comment by edreamleo** [comment]
The outline is not correct after paste-retaining-clones.  Here is the result of the dump-outline command.  As you can see, Node 4 is not properly linked into the outline.  It has no parent field.
```
========================================
----------  <VNode 14805b2e7b8 newHeadline> <===== This is the hidden root node.
len(parents) 0
len(children) 3
parents []
children[ # ===== Correct =====
    <VNode 14806185860 1>,
    <VNode 14809ae0cf8 3>,
    <VNode 14809ab7668 2>
]
----------  <VNode 14806185860 1>
len(parents) 1
len(children) 0
parents [<VNode 14805b2e7b8 newHeadline>]
children[]
----------  <VNode 14809ae0cf8 3>
len(parents) 1
len(children) 1
parents [<VNode 14805b2e7b8 newHeadline>]
children[<VNode 14809ae04a8 4>] # ===== Correct =====
----------  <VNode 14809ae04a8 4>
len(parents) 0
len(children) 0
parents [] # ===== Wrong =====
children[]
----------  <VNode 14809ab7668 2>
len(parents) 1
len(children) 0
parents [<VNode 14805b2e7b8 newHeadline>]
children[]
`

**Comment by edreamleo** [comment]
**SegundoBob**: Perhaps the root of this problem is that Shift-Ctrl-X fails to clear the parent list of the deleted node?

**EKR**: Here is the result of dump-outline after cut-node (Shift-Ctrl-X).  It looks fine.  I strongly suspect the culprit is in paste-retaining clones.
```
========================================
----------  <VNode 1815b05f2e8 newHeadline>
len(parents) 0
len(children) 2
parents []
children[
    <VNode 1815f525278 1>,
    <VNode 1815f1c0898 2>
]
----------  <VNode 1815f525278 1>
len(parents) 1
len(children) 0
parents [<VNode 1815b05f2e8 newHeadline>]
children[]
----------  <VNode 1815f1c0898 2>
len(parents) 1
len(children) 0
parents [<VNode 1815b05f2e8 newHeadline>]
children[]
```

FYI, here is the copied text.  It looks correct to me. That is, the vnodes are properly nested, with the expected gnx's.
```
<?xml version="1.0" encoding="utf-8"?>
<!-- Created by Leo: http://leoeditor.com/leo_toc.html -->
<leo_file xmlns:leo="http://leoeditor.com/namespaces/leo-python-editor/1.1" >
<leo_header file_format="2"/>
<vnodes>
<v t="ekr.20180420164034.1"><vh>3</vh>
<v t="ekr.20180420164036.1"><vh>4</vh></v>
</v>
</vnodes>
<tnodes>
<t tx="ekr.20180420164034.1"></t>
<t tx="ekr.20180420164036.1"></t>
</tnodes>
</leo_file>
```

**Comment by vitalije** [comment]
last night I have tracked down this issue to the leo.core.leoFileCommands.getOutlineFromClipboard method.  Here is the part which I suspect is wrong (or at least I can't understand it fully):
```
    if reassignIndices or tempOutline:
        oldGnxDict = self.gnxDict
        self.gnxDict = {}
    else:
        # All pasted nodes should already have unique gnx's.
        ni = g.app.nodeIndices
        for v in c.all_unique_nodes():
            ni.check_gnx(c, v.fileIndex, v)
```
Perhaps the first `if` line should be:
```
if not reassignIndices or tempOutline:
```
At least changing it fixes this issue. I am not 100% sure this is the correct solution. I don't know what are other uses of this arguments. What is the purpose of tempOutline? Does these two flags differ in any other case? I've got no time to investigate further but this would be my line of thoughts. 

If the above `if` expression is really correct and intended, then the culprit is in cutting the outline. Nodes that were cut off of the outline should be removed from gnxDict before trying to paste them back. The fact that they are still present in the gnxDict makes methods fc.createSaxChildren and fc.createSaxVnode to behave wrongly in this case. It is the outline returned from fc.createSaxChildren that is wrongly linked.

HTH, Vitalije

**Comment by edreamleo** [comment]
On Sat, Apr 21, 2018 at 5:06 AM, Виталије Милошевић <
notifications@github.com> wrote:

> last night I have tracked down this issue to the leo.core.leoFileCommands.
>
getOutlineFromClipboard method.
>
​[snip]​

> Perhaps the first if line should be:
>
> if not reassignIndices or tempOutline:
>
> ​I doubt it. The proposed change seems way too drastic. I would likely
affect regular pastes.  I continue to suspect the paste-retaining-clones
code itself, but I haven't looked at it.

Anyway, it's great to have you back sleuthing with us.

Edward

**Comment by edreamleo** [comment]
On Sat, Apr 21, 2018 at 5:59 AM, Edward K. Ream <edreamleo@gmail.com> wrote:

>
> On Sat, Apr 21, 2018 at 5:06 AM, Виталије Милошевић <
> notifications@github.com> wrote:
>
>> Perhaps the first if line
>> ​ ​
>> ​[of fc.getOutlineFromClipboard] ​
>> ​ should be:​
>> ​
>>
>> if not reassignIndices or tempOutline:
>>
>> ​I doubt it. The proposed change seems way too drastic. I would likely
> affect regular pastes.  I continue to suspect the paste-retaining-clones
> code itself, but I haven't looked at it.
>

​To clarify, paste-retaining-clones simply calls
c.pasteOutline(reassignIndices=False)​.
So by definition, the problem is in c.pasteOutline.

We can also assume c.pasteOutline calls:

    fc.getOutlineFromClipboard(s,
        reassignIndices=False,
        tempOutline=False)

Clearly, the intention is that  c.getOutlineFromClipboard neither reassign
indices nor make entries into the global gnx dict.  Therefore, it looks
like the "if" statement in fc.getOutlineFromClipboard is not the problem.

Instead, the problem must lie somewhere else, that is, code later in
fc.getLeoOutlineFromClipboard. Possibly even in the sax read logic.

Edward

**Comment by edreamleo** [comment]
Dumps cast doubt on this comment and code:
```python
# Important: we must not adjust links when linking v
# into the outline.  The read code has already done that.
if current.hasChildren() and current.isExpanded():
    if check and not self.checkPaste(current, p):
        return None
    p._linkAsNthChild(current, 0, adjust=False)
else:
    if check and not self.checkPaste(current.parent(), p):
        return None
    p._linkAfter(current, adjust=False
```
Indeed, the problem is that p's child does not have a valid proper link.

**Comment by edreamleo** [comment]
Well, it appears that readSaxFile doesn't set parent links properly for *all* descendants of the pasted file.

For example, when paste-retaining-clones pastes the tree: containing 3, 4, 5, resulting in:
```
- 1
- 3
  - 4
    - 5
- 2
```
then neither 4 nor 5 have a parent link.  So this looks like a systematic problem in the sax read code.

**Comment by vitalije** [comment]
This wrong behavior of readSaxFile is present only when a node with the same gnx already exists in gnxDict. If you supply empty gnxDict before calling readSaxFile (or if you remove all nodes that are present in gnxDict, but not connected to the outline), readSaxFile will work properly.

IMO it would be much better if methods like readSaxFile, pasteOutline, pasteOutlineRetainingClones, ... and many others throughout the Leo code, were written to perform just one well defined job. In current implementation they are designed to be universal. So they need several flag arguments that affect their behavior. That makes reading, understanding, debugging and testing such methods much harder than necessary. 

I don't fully understand what all different uses of these methods are. When I looked at this issue the other day, I found that readSaxNode returns well connected outline if gnxDict is clean, i.e. if it doesn't contain any extra node that is not connected to the main outline. That is why I suspected that gnxDict should be temporarily replaced by empty dict. 

Why we wouldn't have two or three versions of the method readSaxFile? Each of them would be straightforward to write and read without flags that affect their behavior. If you are concerned with the repeating code, we have code repetition either way. If the method is universal and take flags as input arguments, then you have to repeatedly write those arguments wherever you use the method. And if this method should call some helper methods, it needs to pass flags again. OTOH if methods are using very similar sequence of lines, then we would indeed have some code repetition, but it would be localized and not spread throughout the code. 

Or, why don't we use method fc.getVnodeFromClipboard that I have added recently to leoFileCommands. I needed this method to perform clean parsing of xml to VNode instance. By *clean* I mean it doesn't disturb gnxDict or any other ivar. There is also method fc.getPosFromClipboard (perhaps the name should be changed) which returns a position of the pasted outline. In effect it returns root position of a temporary created outline which isn't connected to the main outline and whose nodes are not in gnxDict. From there all traversal methods are at hand, and tree can be traversed, transformed in usual way. 

Vitalije

**Comment by edreamleo** [comment]
​​
On Sun, Apr 22, 2018 at 3:43 AM, Виталије Милошевић <
notifications@github.com> wrote:

> This wrong behavior of readSaxFile is present only when a node with the
> same gnx already exists in gnxDict. If you supply empty gnxDict before
> calling readSaxFile (or if you remove all nodes that are present in
> gnxDict, but not connected to the outline), readSaxFile will work properly.
>
​Sure.  That's the behavior when new gnx's are wanted.​

> IMO it would be much better if methods like readSaxFile, pasteOutline,
> pasteOutlineRetainingClones, ... and many others throughout the Leo code,
> were written to perform just one well defined job.
>
​Another bit of advice I'm not likely to forget ;-) It implies significant
restructuring.

First, let me *agree *with you as much as possible:

Recent happy experience with filter.eventFilter and k.masterKeyHandler show
how useful significant restructuring can be. The clone-find commands give
me confidence to undertake refactoring that I would not have considered
previously.

Heh.  Local searches are also useful. readSaxFile does not use its
reassignIndices kwarg!

So I'll consider refactoring readSaxFile while fixing this bug.  But I will
do so only if the code becomes simpler. That's a real possibility,

Second, I'll *disagree *with you in smallish ways:

pasteOutline and pasteOutlineRetainingClones *do* do exactly one thing.
That is, the reassignIndices kwarg to pasteOutline is the one and *only*
difference between the paste-outline and the paste-retaining-clones
commands.

Imo, the reassignIndices kwarg is completely benign. It would violate the
DRY (Don't repeat yourself) principle *not* to use a common helper
(pasteOutline) for the paste-node and paste-retaining-clones commands.

Sure, I could create a new pasteOutlineHelper method, for use by both
commands, but that's a nit. Whatever the helper is called, it *must* have a
"reassignIndices" kwarg.

Maybe other kwargs in pasteOutline could somehow be refactored away, but
this is a small matter. Note that the  undoFlag and redraw flag kwargs came
about, iirc, because they were useful in scripts. Imo, such
"option-oriented" kwargs are benign.

*Summary*

Leo's clone-find commands provide power tools for doing
previously-out-of-the-question refactorings. Refactoring can significantly
improve code, as the "keys2" branch shows.

Imo, keyword args are *not *necessarily a sign that code should be
refactored.  Some are low-cost, low-risk ways of adding minor options to
code.

I'll consider refactoring readSaxFile.  I'll surely remove unused
"reassignIndices" kwarg. Even this minor change implies global changes, for
which the clone-find commands are needed.

Edward

**Comment by edreamleo** [comment]
On Sun, Apr 22, 2018 at 6:00 AM, Edward K. Ream <edreamleo@gmail.com> wrote:

Imo, keyword args are *not *necessarily a sign that code should be
> refactored.  Some are low-cost, low-risk ways of adding minor options to
> code.
>

​There are larger issues in Leo's read/write code to consider.  In
particular, the per-commander fc.gnxDict ​might be a "premature"
optimization. Remembering it complicates getLeoOutlineFromClipboard.

We might be able to dispense with the tempOutline kwarg if all fileCommands
read/write code recomputed the gnxDict as needed.  Of course, this implies
a global change to code, and it also implies a complete scan of the outline
as needed.

So changes are possible, but those changes are dangerous and might have
performance implications.  I am not likely to make them just to remove
kwargs in low-level Leo methods.

Edward

**Comment by vitalije** [comment]
> Imo, keyword args are *not *necessarily a sign that code should be
While I will agree on this, let me add what I think is excessive use of kwarg flags. If it is just one method which perform complete calculation within its definition, then I would say it is o.k. for such method to use kwarg flags. But, if a method performs complicated calculation and uses some helper methods, then I would consider such usage of kwarg flags as excessive. 

For example: pasteOutline and pasteOutlineRetainingClones IMO perform complex operation and so it is not o.k. for them to share same method definition, because pasteOutline passes kwarg flags down to helper methods. It is far better (IMHO) to have them separate definition even if those definition are very similar. After all they change outline in very different way. While one of them is performing clear addition to the outline leaving any already existing node untouched, the other one potentially can change any number of existing nodes. We can say it is different point of view, but the main difference is what method does with input kwarg flags. If it passes them down to helpers, then we have code repetition spreading all over the code. Yes, cff commands can help in future bug fixing and refactoring, but the same command can help refactoring very similar method definition in the same way. Using kwarg flags in these situations sacrifices readability of code to gain exactly what? A smaller code size? If we count how many times those kwarg flags are passed and tested it even may result in larger LOC number. It certainly reduces readability and increases possibility of introducing bugs. If a method has no if statements then it can be tested with just one or two examples of input arguments. Each if statement requires double number of test cases to cover all possible paths of execution. And do we code all those test cases? Most often the answer is no, we don't. 

Ok, pasteOutline and pasteOutlineRetainingClones share some functional parts. They both change outline and both of them must be undoable. But look at the code and you will see that they even compute undodata differently. Even undoFlag is present? So they both can be also not undoable.
Right now in pasteOutline there are 4-5 input kwarg flags and 10 if statements. Theoretically to fully test such method we would need at least 16-32 different combinations of input parameters times 1024 execution paths (thats around 16-32k).

Even if we count more thoroughly and precisely number of required tests we would probably end up with more than 100 required test cases. Are we going to write that number of tests? If OTOH we split that method in several simple and straight implementations with one or zero if statements, we will have some duplicated code, but it may turn out that 4-5 tests would suffice to test everything. 

What if we really write 100 tests for current implementation and later decide to change something. We would need to change 100 tests. That is also violating DRY.

**Comment by edreamleo** [comment]
On Sun, Apr 22, 2018 at 7:54 AM, Виталије Милошевић <
notifications@github.com> wrote:

> For example: pasteOutline and pasteOutlineRetainingClones IMO perform
> complex operation and so it is not o.k. for them to share same method
> definition, because pasteOutline passes kwarg flags down to helper methods.
> It is far better (IMHO) to have them separate definition even if those
> definition are very similar.
>
Imo, c_oc.pasteOutline is fine just the way it is. To my knowledge, it has
never caused any problem, nor is it likely to do so in future.

The paste-retaining-clones bug has *nothing* to do with code style.
Instead, it is a serious logical oversight in fc.getLeoOutlineFromClipboard
or its helper fc.readSaxFile.

Finally, any new unit test will rightly *ignore *the underlying code.
Instead, the unit test will assert that a sequence of operations has an
expected result, namely an outline with expected shape and with expected
links between nodes.

Edward

**Comment by vitalije** [comment]
> The paste-retaining-clones bug has *nothing* to do with code style.
Instead, it is a serious logical oversight in fc.getLeoOutlineFromClipboard
or its helper fc.readSaxFile.

Ok. I don't wont to argue but my point was that these kind of *serious logical oversights* are caused mostly by code style. I took c_oc.pasteOutline just as an example, but there are lot of similar methods throughout Leo code base.

We don't need to agree on our code style preferences, and if at any moment you wish to stop this thread, please let me know.

Let me try again to explain the way I think about this problem. What is the input argument for pasting outline without retaining clones? It is just content of a clipboard and current position. If we really want to use it for both paste with and without retaining clones, it needs also flag argument to decide whether it would reassign indices or not. That makes at most three arguments. However, current implementation has five arguments. So, I question myself why? I can imagine two different explanations. 

1. The function has more than one (or two) purposes that I know of
2. The function has extra input arguments that have no influence on the outcome

Either way I fill that something is bad about that function. If I count how many different execution paths can happen inside this function, that is also very large number. If I look at how many lines of code it has I see it is at least twice as large the maximum recommended size for a function. Well, I know that coding in Leo allows much larger sizes of functions but this particular method doesn't take advantage of Leo outline. It has all lines in a single node. If I look at its helpers and see that they just add more and more complexity I can't help but to feel that it is a mess. Now, you can say it is good as it is now and it worked well for a long time, causing no trouble at all. But it is causing trouble all the time by introducing such a complexity and consuming too much human brain power to understand the code. Developer mind or what is left of it after this big consumption, will likely make a *serious logical oversight*. You may disagree but I feel that the brain power could be used more efficiently if this method hadn't been trying to achieve too many different goals. And this is not the worse defined function in Leo, it is just an example that I have read recently and is related to this issue. 

Another question that I ask myself is: "Why are helpers of this method so far?". I guess it is because the way we store outline on clipboard. Are we using the best possible format for this? JSON come to my mind immediately. If it were JSON format on clipboard, we would not have to use functions for reading outline from a file. You can say that there is no obvious benefit in doing so, but the fact that we are trying to use the same functions that are used for reading xml file is causing additional complexity. Functions that used to have single purpose (reading file and building outline from scratch) now are given another purpose: building  a part of the outline while keeping existing outline nodes untouched. And yet another purpose: to build the part of outline using pre-existing nodes. Now, because of this additional purposes all those methods become more and more complex. That makes modules mutually dependent and so hard to test and reason about. Helpers of c_oc.pasteOutline are located far away and it is hard to follow and understand what is happening inside. If OTOH, we use json module, then all that needs to be done about pasting outline can be located inside the function itself. All helpers can be also defined inside the function. That way if it ever becomes necessary to change this method all changes will be inside and nothing outside can possibly be broken by those changes.

Having helper functions located far away is not a big problem, especially having in mind cff commands. But you wrote somewhere in the Leo documentation about why you like Python, and you gave an example of one liner list comprehension in Python, which would translate in dozens of lines of code in C, and that there was a large number of possible bugs in it. Your conclusion was that this one-liner was by far more readable and preferable way of programming. Now, let me try to apply same logic here.

```
def p2list(p):
    return json.dumps([ p.h, p.b, p.gnx, [p2list(p2) for p2 in p.children()]])
def copyOutline(p):
    g.app.gui.replaceClipboardWith(p2list(p))
```
Paste outline could not be much more complicated than this, and the only difference for retaining clones or not would be in using p.gnx or None instead. Recreating subtree from this recursive list in leoCache module is not trivial. It has to deal with possible differences in clones originated from different files, but for the sake of implementing pasteOutline command those differences can be safely ignored and all existing nodes simply replaced with the version from the clipboard. Pasting without retaining clones is even simpler.

**Comment by edreamleo** [comment]
On Sun, Apr 22, 2018 at 2:25 PM, Виталије Милошевић <
notifications@github.com> wrote:

my point was that these kind of *serious logical oversights* are caused
> mostly by code style.


​It's time to actually fix this bug ;-)
​


> We don't need to agree on our code style preferences, and if at any moment
> you wish to stop this thread, please let me know.


​Several good things will come from our discussion.​


​First, we get to explain our points of view more fully :-)

Second, we can shift from a general discussion of principles to review of
particular methods.​  Some of the kwargs in c_oc.pasteOutline might be
removed. It's *always* good to simplify code.

Third, some methods, like at.read, might indeed be further simplified.  I
would welcome suggestions.  We haven't ever conducted a formal code review
of any method, and I think some methods would benefit from a review.

> Let me try again to explain the way I think about this problem. What is
> the input argument for pasting outline without retaining clones? It is just
> content of a clipboard and current position. If we really want to use it
> for both paste with and without retaining clones, it needs also flag
> argument to decide whether it would reassign indices or not. That makes at
> most three arguments. However, current implementation has five arguments.
> So, I question myself why?
>
​As I said above, I wonder about that myself.  cff will explain. The first
hit is in abbrev.paste_tree:

    p = c.pasteOutline(s=s, redrawFlag=False, undoFlag=False)

So there is a reason for the args, even though I'm not really sure why the
abbreviation code wants to set the flags as it does.

Well, couldn't we do as you suggest, and remove some funtionality from
c.pasteOutline?  Yes, we could, and perhaps the result would be better.

In short, a code review is reasonable.
​


> Either way I fill that something is bad about that function.


​Imo, you are overstating the harm that the extra kwargs cause.​

​ The redraw flag has a clear meaning, as does the undoFlag.  There can be
no "exponential explosion" of execution paths.  There is, in fact, one
path, with minor digressions.​

> Developer mind or what is left of it after this big consumption, will
> likely make a *serious logical oversight*. You may disagree but I feel
> that the brain power could be used more efficiently if this method hadn't
> been trying to achieve too many different goals.
>
​I strongly disagree.  In fact, Leo already has what you want, namely code
that is easily maintained.  It is easy to find the code that does any
particular job, and there are minimal dependencies between different pieces
of code.

> Another question that I ask myself is: "Why are helpers of this method so
> far?". I guess it is because the way we store outline on clipboard. Are we
> using the best possible format for this? JSON come to my mind immediately.
>
​This is a completely separate discussion.  Leo puts things to the
clipboard in xml format because that's the format of .leo files. As a
result, the read code is a minor variant of the code that reads external
files.  Furthermore, it's possible to paste a node into an external editor
and save the pasted text as a .leo file.  That would not be possible if
json were used.

> If OTOH, we use json module, then all that needs to be done about pasting
> outline can be located inside the function itself.
>
​Well, perhaps Leo should use json as a file format.  It might simplify
some code, but again, this is a completely separate topic.​

> All helpers can be also defined inside the function. That way if it ever
> becomes necessary to change this method all changes will be inside and
> nothing outside can possibly be broken by those changes.
>
​I dislike this style.  I much prefer having helper functions be child
nodes of the code that uses the helpers.​

> you wrote somewhere in the Leo documentation about why you like Python,
> and you gave an example of one liner list comprehension in Python, which
> would translate in dozens of lines of code in C, and that there was a large
> number of possible bugs in it. Your conclusion was that this one-liner was
> by far more readable and preferable way of programming. Now, let me try to
> apply same logic here.
>
> def p2list(p):
>     return json.dumps([ p.h, p.b, p.gnx, [p2list(p2) for p2 in p.children()]])
> def copyOutline(p):
>     g.app.gui.replaceClipboardWith(p2list(p))
>
>
​Yes, this might be better, but this not *just* a style-related question.


> Paste outline could not be much more complicated than this, and the only
> difference for retaining clones or not would be in using p.gnx or None
> instead.


​The code that copies nodes to the clipboard must *always* write gnx's.  It
is only the read logic that must choose whether to retain or reallocate
gnx's.
​


> Recreating subtree from this recursive list in leoCache module is not
> trivial. It has to deal with possible differences in clones originated from
> different files, but for the sake of implementing pasteOutline command
> those differences can be safely ignored and all existing nodes simply
> replaced with the version from the clipboard. Pasting without retaining
> clones is even simpler.


​paste-retaining clones must use *existing* vnodes when pasting, if those
nodes exist *anywhere* in the outline.  As you say, this is not a trivial
task. It's a global problem.  ​
​readSaxFile and helpers do this. Similar logic would be required if we
used json instead.

Maintaining clone links is just hard, and has little to do with data
representation.  The fix is likely to be to add one or two lines of code
somewhere.

*Summary*

It's always good to review code with fresh eyes.  Seemingly minor
simplifications can cascade into code collapse. I welcome specific
suggestions. We might start with at.read.

paste-retaining-clones is much more difficult than a simple paste, because
all vnodes with the same gnx *must* be the same vnode.  As a result, this
code will never be easy.

This bug is not a result of poor coding practices.  It's just a mistake.

Edward

**Comment by vitalije** [comment]
I have edited my last post and it seems that you didn't read the latest version of it. 
Here is a full script that copies selected node using json format.
```
import json
def p2list(p):
    return [ p.h, p.b, p.gnx, [p2list(p2) for p2 in p.children()]]
def copyOutline(p):
    g.app.gui.replaceClipboardWith(json.dumps(p2list(p)))
copyOutline(p)
```
you can paste this code in a single node and make a script button of it. Then select some outline and execute it to copy outline in json format. 

And here is a script that will paste the outline retaining clones.
```
import json
def createOutline(parent_v, aList, gnxDict):
    h, b, gnx, children = aList
    if h is not None:
        parent_v._headString = h
        parent_v._bodyString = b
    for child_tuple in children:
        h, b, gnx, grandChildren = child_tuple
        if gnx in gnxDict:
            child_v = gnxDict[gnx]
            del child_v.children[:]
        else:
            child_v = leoNodes.VNode(c, gnx)
            gnxDict[gnx] = child_v
        if parent_v not in child_v.parents:
            child_v.parents.append(parent_v)
        parent_v.children.append(child_v)
        createOutline(child_v, child_tuple, gnxDict)

def pasteRetainingClones(p):
    aList = json.loads(g.app.gui.getTextFromClipboard())
    v1 = c.fileCommands.gnxDict.get(aList[2])
    if v1 is None:
        v1 = leoNodes.VNode(c, aList[2])
    else:
        del v1.children[:]
    p.v.children.append(v1)
    if not p.v in v1.parents:
        v1.parents.append(p.v)
    createOutline(v1, aList, c.fileCommands.gnxDict)
    c.redraw()
pasteRetainingClones(p)
```
For a paste without retaining clones here is additional function
```
import json
def clearGnxes(aList):
    aList[2] = None
    for xl in aList[3]:
        clearGnxes(xl)

def pasteOutline(p):
    aList = json.loads(g.app.gui.getTextClipboard())
    clearGnxes(aList)
    pasteRetainingClones(p)
pasteOutline(p)
```
Add some guards to check if the text on clipboard is in proper format  and support for undo. This whole script is less than 50LOC. With support for undo and some checks it can grow to 100LOC. Compare it with the present code.

>> All helpers can be also defined inside the function. That way if it ever
 becomes necessary to change this method all changes will be inside and
 nothing outside can possibly be broken by those changes.

> ​I dislike this style.  I much prefer having helper functions be child
nodes of the code that uses the helpers.​

I haven't said that helper functions need to be in the same node. They certainly should be in separate nodes located as child nodes of the function that uses them as you like it. My point was just to add at-others line inside the top level function. That way all helpers are invisible for the rest of code. You can change them anyway you like and they will never cause any other function to fail. If some helper appears often enough than it may deserve to be defined outside as a standalone function. All helpers share input arguments of the main function so this arguments need not to be passed everywhere. 

> This bug is not a result of poor coding practices.  It's just a mistake.

It never is. Every single bug in the end is just a mistake. Even if one uses the best possible coding style it is not guaranteed that he won't make a mistake. Coding style is just about probability of making a mistake. Probability is a gray area. I can't prove that the probability of making mistakes would be considerable less if the coding style that I propose is accepted. But that is what I believe to be true. OTOH I don't suppose that you could prove that this probability would not go down either.

Now, about `at.read` code. I have made some experiments last summer and the code is in leo-snippets 
https://github.com/leo-editor/snippets/tree/master/experiments

AFAIR read function successfully reads external files with sentinels. At least all external files that are part of Leo distribution (or were part at the time of writing this tests). The files are successfully read and compared with the outline that at.read creates.

There is also write function that writes external files and it is now part of line-numbering plugin.

You can try and test them yourself. They are completely self-contained, and can be attached to commander instance. AFAIR no ivars of any other module are disturbed by them so I don't expect any clash with the existing code.

These methods could be even simpler if support for some less used directives was dropped. I don't remember which exactly. I remember that when I rewrote this functions in Rust and clojurescript/coffeescript I realized that considerable simplification can be achieved if some directives that I personally have never used and don't know if anyone else is using them are eliminated. But I don't know which and why. When the time comes I can look again and refresh my memory.

Anyway even without those possible simplifications, the functions I wrote last summer support all Leo directives and read/write all Leo's files the same way Leo currently does.

**Comment by edreamleo** [comment]
Rev 6d05a6ee0 in devel merges the paste-retaining-clones branch into devel.

This should close this issue. Please reopen if there are problems.

I have hidden all comments pertaining to programming style.  We'll continue this discussion on leo-editor.
